src/bitmap/bytes.rs

```diff
@@ -0,0 +1,255 @@
+///
+/// This bitmap requires `O(n)` space and can be read and wrote to in `O(1)` time.
+///
+/// This type is fast for both reads and writes, trading internal complexity for speed.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct BytesBitmap {
+    max_key: usize,
+    bitmap: BytesMut,
+}
+
+impl BytesBitmap {
+    pub(crate) fn into_parts(self) -> (Vec<usize>, usize) {
+        let mut vec = Vec::with_capacity(self.bitmap.len() / size_of::<usize>());
+        let chunks = self.bitmap.chunks_exact(size_of::<usize>());
+        for chunk in chunks {
+            vec.push(usize::from_ne_bytes(chunk.try_into().unwrap()));
+        }
+        (vec, self.max_key)
+    }
+
+    pub(crate) fn shrink(&mut self) {
+        self.bitmap.resize(self.bitmap.len(), 0);
+    }
+
+    pub fn bitmap(self) -> Bytes {
+        self.bitmap.freeze()
+    }
+
+    pub fn max_key(&self) -> usize {
+        self.max_key
+    }
+
+    pub fn from_bytes(bitmap: Bytes, max_key: usize) -> Self {
+        Self {
+            max_key,
+            bitmap: BytesMut::from(bitmap),
+        }
+    }
+}
+
+impl Bitmap for BytesBitmap {
+    fn new_with_capacity(max_key: usize) -> Self {
+        let size = (index_for_key(max_key) + 1) * size_of::<usize>();
+        let mut bytes = BytesMut::zeroed(size);
+
+        Self {
+            bitmap: bytes,
+            max_key,
+        }
+    }
+
+    fn set(&mut self, key: usize, value: bool) {
+        let offset = index_for_key(key);
+        let byte_offset = offset * size_of::<usize>();
+
+        let slice = &mut self.bitmap[byte_offset..byte_offset + size_of::<usize>()];
+        let mut num = usize::from_ne_bytes(slice.try_into().unwrap());
+
+        if value {
+            num |= bitmask_for_key(key);
+        } else {
+            num &= !bitmask_for_key(key);
+        }
+
+        slice.copy_from_slice(&num.to_ne_bytes());
+    }
+
+    fn get(&self, key: usize) -> bool {
+        let offset = index_for_key(key);
+        let byte_offset = offset * size_of::<usize>();
+        let slice = &self.bitmap[byte_offset..byte_offset + size_of::<usize>()];
+        let num = usize::from_ne_bytes(slice.try_into().unwrap());
+        num & bitmask_for_key(key) != 0
+    }
+
+    fn byte_size(&self) -> usize {
+        self.bitmap.len()
+    }
+
+    fn or(&self, other: &Self) -> Self {
+        assert_eq!(self.bitmap.len(), other.bitmap.len());
+
+        let mut result = BytesMut::with_capacity(self.bitmap.len());
+        let chunks = self
+            .bitmap
+            .chunks_exact(size_of::<usize>())
+            .zip(other.bitmap.chunks_exact(size_of::<usize>()));
+
+        for (a_chunk, b_chunk) in chunks {
+            let a = usize::from_ne_bytes(a_chunk.try_into().unwrap());
+            let b = usize::from_ne_bytes(b_chunk.try_into().unwrap());
+            result.put_slice(&(a | b).to_ne_bytes());
```

Comment:
so i tried a few variations of this and i can't get it to consistently pass the tests.

for example:
```
// fails the tests
for i in 0..self.bitmap.len() {
    result.put_u8(self.bitmap[i] | other.bitmap[i]);
}

// won't compile due to no language support for `&[u8] | &[u8]`
for (a, b) in self.bitmap.iter().zip(other.bitmap.iter()) {
    result.put_u8(a | b);
}

// fails tests
result.extend_from_slice(&self.bitmap.iter()
    .zip(other.bitmap.iter())
    .map(|(a, b)| a | b)
    .collect::<BytesMut>());
```

i'm like 85% sure llvm is going to auto-vectorize this a bit because the iterators guaranteed to be ordered and the same size (due to the assertion). i'm also fairly positive the casts are `noop` because they're realistically just compiler hints. the only other thing i could really do would be to add SIMD intrinsics but those are inherently `unsafe`. i added the `#[inline]` hint just in case llvm needs it but i'm not sure this can be optimized further.

if you have other ideas i'm open to them.